src/ConnectionAutomaton.cpp

```diff
@@ -241,15 +297,44 @@ void ConnectionAutomaton::cancel() {
 }
 /// @}
 
+void ConnectionAutomaton::scheduleKeepalive(const uint32_t keepAliveDelay) {
+  if (keepAliveDelay > 0) {
+    auto eventBase = folly::EventBaseManager::get()->getExistingEventBase();
+    CHECK(eventBase);
+
+    eventBase->runAfterDelay(
+        [this, keepAliveDelay]() { sendKeepalive(keepAliveDelay); },
+        keepAliveDelay);
+  }
+}
+
+void ConnectionAutomaton::sendKeepalive(const uint32_t keepAliveDelay) {
+  // TODO is this check safe? or needs to check a synchronized shared flag?
```

Comment:
since we are assuming everything runs on the same EventBase, I believe you are safe not to synchronize.